src/utils/dockerDirManager.ts

```diff
@@ -0,0 +1,303 @@
+    try {
+      const rawConfig = await fs.promises.readFile(this.dockerConfigPath, { encoding: 'utf-8' });
+
+      return JSON.parse(rawConfig);
+    } catch (error: any) {
+      if (error.code !== 'ENOENT') {
+        throw error;
+      }
+
+      return {};
+    }
+  }
+
+  /**
+   * Writes the docker CLI config.json file.
+   * @param config An object that is the config we want to write.
+   */
+  protected async writeDockerConfig(config: PartialDockerConfig): Promise<void> {
+    const rawConfig = JSON.stringify(config);
+
+    await fs.promises.mkdir(this.dockerDirPath, { recursive: true });
+    await fs.promises.writeFile(this.dockerConfigPath, rawConfig, { encoding: 'utf-8' });
+  }
+
+  /**
+   * Read the docker configuration, and return the docker socket in use by the
+   * current context.  If the context is invalid, return the default socket
+   * location.
+   * @param currentContext Docker's current context, as set in the configs.
+   */
+  protected async getCurrentDockerSocket(currentContext?: string): Promise<string> {
+    const defaultSocket = `unix://${ this.defaultDockerSockPath }`;
+
+    if (!currentContext) {
+      return defaultSocket;
+    }
+
+    for (const dir of await fs.promises.readdir(this.dockerContextDirPath)) {
+      const contextPath = path.join(this.dockerContextDirPath, dir, 'meta.json');
+
+      try {
+        const data = yaml.parse(await fs.promises.readFile(contextPath, 'utf-8'));
+
+        if (data.Name === currentContext) {
+          return data.Endpoints?.docker?.Host as string ?? defaultSocket;
+        }
+      } catch (ex) {
+        console.log(`Failed to read context ${ dir }, skipping: ${ ex }`);
+      }
+    }
+
+    // If we reach here, the current context is invalid.
+    return defaultSocket;
+  }
+
+  /**
+   * Given some information about state external to this method, returns the
+   * name of the context that should be used. Follows these rules, in order of preference:
+   * 1. If we have control of the default socket (`/var/run/docker.sock`), we should set the
+   *    context to it (which is actually un-setting the `currentContext` key).
+   *    This should have the widest compatibility.
+   * 2. Otherwise, check the current context and don't change anything if any of the following
+   *    is true:
+   *    - The current context uses a valid unix socket - the user is probably using it.
+   *    - The current context uses a non-unix socket (e.g. tcp) - we can't check if it's valid.
+   * 3. The current context is invalid - set the current context to our (rancher-desktop) context.
+   * @param weOwnDefaultSocket Whether Rancher Desktop has control over the default socket.
+   * @param currentContext The current context.
+   * @returns Undefined for default context; string containing context name for other contexts.
+   */
+  async getDesiredDockerContext(weOwnDefaultSocket: boolean, currentContext: string | undefined): Promise<string | undefined> {
```

Comment:
This can be `protected`?

Reply:
I had it `protected` originally, but changed it because I needed to call it in unit tests. Are you aware of a way of getting around this? It would be great to know about

Reply:
I believe that if you use `obj['foo']` instead of `obj.foo`, you can work around the TypeScript protections. (Yes, that is intentional, and we shouldn't abuse it outside of tests.)

This doesn't help with the `#foo` private members that's JavaScript-native.

Reply:
Ah, that's what that was for.

Also, something you might be interested in (if you don't know about it already):

```
const getCurrentDockerSocketMock = jest.spyOn(subj as any, 'getCurrentDockerSocket')
    .mockResolvedValue('some-url');
```

The `subj as any` lets us get around the whole `protected` thing, but for mocks. Very useful.